### PR TITLE
Add input vector for association metrics

### PR DESCRIPTION
**Pull request by DiveFish**
Query association strength between two tokens for any possible dependency relation.

So far, input layers only stored the `i32` indices that were lookups into an embedding matrix or a lookup table. A new kind of input vector is now added which makes it possible to use `f32` values directly as an input vector. This is convenient for e.g. including association measures like PMI into the training process of the parser. PMIs are retrieved for the parser states that are involved in an attachment. The lookup of the PMIs is currently done via a HashMap which can be fed with PMIs from a file.

### CODE REVIEW

**Comment by DiveFish** [comment]
This PR implements the changes requested in the PR on the danieldk/dpar original repo (https://github.com/danieldk/dpar/pull/32) and is meant to be merged into the SFB repo. PR https://github.com/danieldk/dpar/pull/32 will be left open until this PR in the SFB repo is closed so that all comments made so far remain visible.

File: dpar-utils/src/bin/dpar-train.rs [hunk 1]
```diff
@@ -65,68 +66,78 @@ fn main() {
         FileProgress::new(input_file).or_exit("Cannot create progress bar", 1),
     ));
     eprintln!("Vectorizing training data...");
-    let (train_labels, train_inputs) =
+    let (train_labels, train_inputs, train_metric_inputs) =
```
Comment (danieldk) [inline dpar-utils/src/bin/dpar-train.rs]: This still uses the term 'metric inputs'.

File: dpar-utils/src/bin/dpar-train.rs [hunk 2]
```diff
@@ -65,68 +66,78 @@ fn main() {
...
+    validation_metric_inputs: Vec<Tensor<f32>>,
 ) -> Result<(), Error>
 where
     S: SerializableTransitionSystem,
 {
     let system = S::default();
     let lookups = config.lookups.load_lookups()?;
     let inputs = config.parser.load_inputs()?;
-    let vectorizer = InputVectorizer::new(lookups, inputs);
+    let association_strengths = HashMap::new();
```
Comment (danieldk) [inline dpar-utils/src/bin/dpar-train.rs]: Move construction to the main function, pass as argument to `train_with_system`, etc.

File: dpar-utils/src/bin/dpar-train.rs [hunk 3]
```diff
@@ -209,11 +229,13 @@ where
         ProgressStyle::default_bar()
             .template(&format!("{{bar}} {} batch {{pos}}/{{len}}", epoch_type)),
     );
-    for (labels, inputs) in labels.iter().zip(inputs.iter()) {
+    for (labels, (inputs, metric_inputs)) in
```
Comment (danieldk) [inline dpar-utils/src/bin/dpar-train.rs]: Consider the `izip!` macro from `itertools`.

File: dpar-utils/src/bin/dpar-train.rs [hunk 4]
```diff
@@ -254,14 +276,15 @@ where
...
-) -> Result<(Vec<Tensor<i32>>, Vec<LayerTensors<i32>>), Error>
+) -> Result<(Vec<Tensor<i32>>, Vec<LayerTensors<i32>>, Vec<Tensor<f32>>), Error>
 where
     R: BufRead,
     S: SerializableTransitionSystem,
 {
     let lookups = config.lookups.load_lookups()?;
     let inputs = config.parser.load_inputs()?;
-    let vectorizer = InputVectorizer::new(lookups, inputs);
+    let association_strengths = HashMap::new();
```
Comment (danieldk) [inline dpar-utils/src/bin/dpar-train.rs]: Construct `HashMap` in `main` and pass it to `collect_with_system`.

File: dpar-utils/src/stored_table.rs [hunk 5]
```diff
@@ -76,6 +76,10 @@ impl Lookup for StoredLookupTable {
         }
     }
 
+    fn lookup_values(&self, values: &mut Vec<String>) {
+        //TODO: Implement
```
Comment (danieldk) [inline dpar-utils/src/stored_table.rs]: Do not add unimplemeted functions. Or if it is absolutely necessary, use the `unimplemented` macro so that the code panics when a codepath reaching it is triggered:

https://doc.rust-lang.org/std/macro.unimplemented.html

File: dpar/src/features/input_layers.rs [hunk 6]
```diff
@@ -166,21 +178,53 @@ impl InputVectorizer {
...
     }
 
     /// Vectorize a parser state into the given slices.
-    pub fn realize_into<S>(&self, state: &ParserState, slices: &mut EnumMap<Layer, S>)
-    where
+    pub fn realize_into<S>(
+        &self,
+        state: &ParserState,
+        lookup_slices: &mut EnumMap<Layer, S>,
+        non_lookup_slices: &mut [f32],
```
Comment (danieldk) [inline dpar/src/features/input_layers.rs]: `s/non_lookup_slices/non_lookup_slice/g`

File: dpar/src/features/input_layers.rs [hunk 7]
```diff
@@ -227,6 +271,89 @@ impl InputVectorizer {
...
+    ///
+    /// Add to `non_lookup_slices` the association measure between all parser state addresses
+    /// undergoing attachment.
+    pub fn realize_into_assoc_strengths(
+        &self,
+        state: &ParserState,
+        non_lookup_slices: &mut [f32],
+        attachment_addrs: &[AttachmentAddr],
+    ) {
+        for (addr, non_lookup_slice) in attachment_addrs.iter().zip(non_lookup_slices.iter_mut()) {
```
Comment (danieldk) [inline dpar/src/features/input_layers.rs]: These are zipped, however, as far as I can see, they are not the same length? `attachment_addrs` has length 2, `non_lookup_slices` has the number of dependency relations times 2. You probably want to use a `repeat` iterator for `attachment_addrs`.

File: dpar/src/features/input_layers.rs [hunk 8]
```diff
@@ -227,6 +271,89 @@ impl InputVectorizer {
...
+            };
+            let addr_dependent = addr::AddressedValue {
+                address: vec![addr.dependent],
+                layer: addr::Layer::Token,
+            };
+
+            let head = addr_head.get(state);
+            let dependent = addr_dependent.get(state);
+
+            if head.is_some()
```
Comment (danieldk) [inline dpar/src/features/input_layers.rs]: Use a `match` or `if let` expression. Along the lines of:

```
if let  (Some(head), Some(dependent), Some(layer)) = (head, dependent, self.layer_lookups.layer_lookup(Layer::DepRel) {
  // ...
}
```

allows you to check that they are all `Some`, while simultaneously unpacking the wrapped values :).

File: dpar/src/features/input_layers.rs [hunk 9]
```diff
@@ -227,6 +271,89 @@ impl InputVectorizer {
...
+                && dependent.is_some()
+                && self.layer_lookups.layer_lookup(Layer::DepRel).is_some()
+            {
+                let head = head.unwrap().to_owned();
+                let dependent = dependent.unwrap().to_owned();
+
+                let mut deprels = Vec::new();
+                self.layer_lookups
```
Comment (danieldk) [inline dpar/src/features/input_layers.rs]: I am not sure what this does? I think we discussed adding something to lookup that allows you to get/iterate over all keys in the lookup.

File: dpar/src/features/input_layers.rs [hunk 10]
```diff
@@ -227,6 +271,89 @@ impl InputVectorizer {
...
+                    .layer_lookup(Layer::DepRel)
+                    .unwrap()
+                    .lookup_values(&mut deprels);
+
+                for deprel in deprels {
```
Comment (danieldk) [inline dpar/src/features/input_layers.rs]: This is overwriting the same element of `non_lookup_slices` over and over again.

File: dpar/src/features/input_layers.rs [hunk 11]
```diff
@@ -227,6 +271,89 @@ impl InputVectorizer {
...
+                }
+            }
+        }
+    }
+
+    fn assoc_strength(&self, head: &str, dependent: &str, deprel: &str) -> f32 {
+        let dep_triple = (head.to_string(), dependent.to_string(), deprel.to_string());
+        match self.association_strengths.get(&dep_triple) {
+            Some(association_strength) => association_strength.to_owned(),
+            None => 0.0.to_owned()
```
Comment (danieldk) [inline dpar/src/features/input_layers.rs]: Something of later: but I guess that the default may be measure-specific (e.g. 0.5 for probabilities).

File: dpar/src/features/input_layers.rs [hunk 12]
```diff
@@ -227,6 +271,89 @@ impl InputVectorizer {
...
+        }
+    }
+
+    /// Read association strengths for dependency triples from a text file.
```
Comment (danieldk) [inline dpar/src/features/input_layers.rs]: Does not belong here, should probably be in dpar-utils.

File: dpar/src/features/lookup.rs [hunk 13]
```diff
@@ -56,6 +59,10 @@ impl Lookup for Embeddings {
         self.indices().get(feature).cloned()
     }
 
+    fn lookup_values(&self, values: &mut Vec<String>) {
```
Comment (danieldk) [inline dpar/src/features/lookup.rs]: Return `&[String]`, much cheaper.

File: dpar/src/features/lookup.rs [hunk 14]
```diff
@@ -91,6 +98,10 @@ impl Lookup for MutableLookupTable {
         Some(numberer.add(feature.to_owned()))
     }
 
+    fn lookup_values(&self, values: &mut Vec<String>) {
```
Comment (danieldk) [inline dpar/src/features/lookup.rs]: Same here, though the `borrow` may make things a bit ugly.

File: dpar/src/features/input_layers.rs (hunk 10 above)
Comment (danieldk) [reply dpar/src/features/input_layers.rs]: It seems that the unit test is incomplete, since it does not catch these errors?

File: dpar/src/models/tensorflow/collector.rs [hunk 15]
```diff
@@ -95,10 +114,22 @@ where
...
+            .unwrap()
+            .len();
+
         if self.instance_idx == 0 {
             let layer_tensors = self.new_layer_tensors(self.batch_size);
-            self.inputs.push(layer_tensors);
+            self.lookup_inputs.push(layer_tensors);
             self.labels.push(Tensor::new(&[self.batch_size as u64]));
+            self.non_lookup_inputs
+                .push(Tensor::new(&[
```
Comment (danieldk) [inline dpar/src/models/tensorflow/collector.rs]: Shouldn't this be:

`&[self.batch_size, n_deprels * T::ATTACHMENT_ADDRS.len()]`

?

File: dpar/src/models/tensorflow/collector.rs [hunk 16]
```diff
@@ -108,7 +139,15 @@ where
 
         self.vectorizer.realize_into(
             state,
-            &mut self.inputs[batch].to_instance_slices(self.instance_idx),
+            &mut self.lookup_inputs[batch].to_instance_slices(self.instance_idx),
+            &mut self.non_lookup_inputs[batch][(self.instance_idx
```
Comment (danieldk) [inline dpar/src/models/tensorflow/collector.rs]: Factor out the shared parts of the range bounds to a varuable. (Something like `non_lookup_inputs_offset`.)

**Review by danieldk — CHANGES_REQUESTED** [review]

File: dpar/src/features/lookup.rs (hunk 14 above)
Comment (DiveFish) [reply dpar/src/features/lookup.rs]: Yes, that is why I went for the detour of modifying a vec that's passed as an argument

File: dpar-utils/src/bin/dpar-train.rs (hunk 1 above)
Comment (DiveFish) [reply dpar-utils/src/bin/dpar-train.rs]: Thanks for the pointer!

File: dpar/src/features/lookup.rs (hunk 14 above)
Comment (danieldk) [reply dpar/src/features/lookup.rs]: The `borrow` is annoying, but in these cases you can return a `Cow` value.

https://doc.rust-lang.org/std/borrow/enum.Cow.html

Return a borrowed copy when you can, otherwise return an owned value.

File: dpar/src/features/input_layers.rs (hunk 7 above)
Comment (DiveFish) [reply dpar/src/features/input_layers.rs]: Yes, you're right, the lengths are different. I'll adjust this.

File: dpar/src/features/input_layers.rs (hunk 8 above)
Comment (DiveFish) [reply dpar/src/features/input_layers.rs]: Much nicer!

File: dpar/src/features/input_layers.rs (hunk 9 above)
Comment (DiveFish) [reply dpar/src/features/input_layers.rs]: Yes, it gets the list of all dependency relations as a `Vec<String>`

File: dpar/src/features/input_layers.rs (hunk 10 above)
Comment (DiveFish) [reply dpar/src/features/input_layers.rs]: Probably cause there is only one dependency relation in the test which doesn't trigger overwriting. I will add a test with more than 1 deprel.

File: dpar/src/features/input_layers.rs (hunk 12 above)
Comment (DiveFish) [reply dpar/src/features/input_layers.rs]: Or just put it into a `util.rs` within `dpar/features`, as it's been done for the `dep_head_mapping(...)` method in `systems/util.rs`?

**Comment by danieldk** [comment]
Git pro-tip: rebase your branch on master rather than merging. Then you do not have these pesky merge commits that you need to get rid of later ;). And it's usually the same amount of work (need to fix merge conflicts in either case).

File: dpar-utils/src/stored_table.rs [hunk 17]
```diff
@@ -76,6 +77,13 @@ impl Lookup for StoredLookupTable {
         }
     }
 
+    fn lookup_values<'a>(&'a self) -> Cow<'a, Vec<String>> {
```
Comment (danieldk) [inline dpar-utils/src/stored_table.rs]: `Cow<'a, [String]>`

**Review by danieldk — CHANGES_REQUESTED** [review]

**Comment by DiveFish** [comment]
A lesson in rebasing vs. merging 😅  The changes should be complete now
